README.md
<h1 align="center">Rehype Cloudinary Docusaurus plugin</h1>

<p align="center">Cloudinary offers an image CDN which can improve performance of your site. This plugin allows Docusaurus to use Cloudinary to serve optimised images.</p>

<p align="center">
	<a href="https://www.npmjs.com/package/rehype-cloudinary-docusaurus" target="_blank">
		<img alt="rehype-cloudinary-docusaurus version on npm" src="https://img.shields.io/npm/v/rehype-cloudinary-docusaurus.svg" />
	</a>
	<a href="https://img.shields.io/npm/dm/rehype-cloudinary-docusaurus.svg" target="_blank">
		<img alt="Downloads on npm" src="https://img.shields.io/npm/dm/rehype-cloudinary-docusaurus.svg" />
	</a>
	<a href="https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/actions/workflows/build.yml" target="_blank">
		<img alt="Build" src="https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/actions/workflows/build.yml/badge.svg" />
	</a>
	<a href="https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/blob/main/.github/CODE_OF_CONDUCT.md" target="_blank">
		<img alt="Contributor Covenant" src="https://img.shields.io/badge/code_of_conduct-enforced-21bb42" />
	</a>
	<a href="https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/blob/main/LICENSE.md" target="_blank">
	    <img alt="License: MIT" src="https://img.shields.io/github/license/johnnyreilly/rehype-cloudinary-docusaurus?color=21bb42">
    </a>
	<a href="https://github.com/sponsors/johnnyreilly" target="_blank">
    	<img alt="Sponsor: On GitHub" src="https://img.shields.io/badge/sponsor-on_github-21bb42.svg" />
    </a>
	<img alt="Style: Prettier" src="https://img.shields.io/badge/style-prettier-21bb42.svg" />
    <img alt="TypeScript: Strict" src="https://img.shields.io/badge/typescript-strict-21bb42.svg" />
</p>

## Usage

```shell
npm i rehype-cloudinary-docusaurus
```

We need to reference it in our `docusaurus.config.js` file. We do this by adding it to the `rehypePlugins` array:

```js
//@ts-check
import docusaurusCloudinaryRehypePlugin from "rehype-cloudinary-docusaurus";

const url = "https://johnnyreilly.com";

/** @type {import('@docusaurus/types').Config} */
const config = {
	// ...
	presets: [
		[
			"@docusaurus/preset-classic",
			/** @type {import('@docusaurus/preset-classic').Options} */
			({
				// ...
				blog: {
					// ...
					rehypePlugins: [
						[
							rehypeCloudinaryDocusaurus,
							{
								cloudName: "demo",
								baseUrl: url,
							},
						],
					],
					// ...
				},
				// ...
			}),
		],
	],
	// ...
};
```

You will also need to disable the `url-loader` in your Docusaurus build which transforms images into base64 strings, as this will conflict with the plugin. [There isn't a first class way to do this in Docusaurus at present](https://github.com/facebook/docusaurus/pull/5498). However by setting the environment variable `WEBPACK_URL_LOADER_LIMIT` to `0` you can disable it. [You can see an implementation example in this pull request](https://github.com/johnnyreilly/blog.johnnyreilly.com/pull/397). It amounts to adding the [`cross-env`](https://github.com/kentcdodds/cross-env) package and then adding the following to your `package.json`:

```json
	"build": "cross-env WEBPACK_URL_LOADER_LIMIT=0 docusaurus build",
```

See this package in action on: <https://johnnyreilly.com> - the source code can be found here: <https://github.com/johnnyreilly/blog.johnnyreilly.com>. To read more on how this plugin came to be, see [this blog post](https://johnnyreilly.com/docusaurus-image-cloudinary-rehype-plugin) and [how it was migrated to MDX 3](https://johnnyreilly.com/docusaurus-3-how-to-migrate-rehype-plugins).

## Development

See [`.github/CONTRIBUTING.md`](./.github/CONTRIBUTING.md).
Thanks! ❤️🌻

> 💙 This package is based on [@JoshuaKGoldberg](https://github.com/JoshuaKGoldberg)'s [template-typescript-node-package](https://github.com/JoshuaKGoldberg/template-typescript-node-package).


CHANGELOG.md
# Changelog

## [3.0.0](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/compare/1.3.1...3.0.0) (2023-10-27)

### ⚠ BREAKING CHANGES

- docusaurus v3 version (#9)

### Features

- docusaurus v3 version ([#9](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/issues/9)) ([d48d9fe](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/commit/d48d9fe67bdb9f3d16525b3de30d8dcc51672e11))

### [1.3.1](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/compare/1.3.0...1.3.1) (2023-03-01)

## [1.3.0](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/compare/1.2.1...1.3.0) (2023-01-28)

### Features

- add w_auto,dpr_auto support ([#7](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/issues/7)) ([ab66673](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/commit/ab66673d826167f6403103fc937b347b30e5381a))

### [1.2.1](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/compare/1.2.0...1.2.1) (2023-01-18)

### Bug Fixes

- remark -> rehype rename ([#6](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/issues/6)) ([ab763fa](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/commit/ab763fa88860723a881fc4d54e1b1a333d105e2f))

## [1.2.0](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/compare/1.1.0...1.2.0) (2023-01-13)

### Features

- use f_auto,q_auto ([#5](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/issues/5)) ([6aa6492](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/commit/6aa6492a974d464773fca424057fcfda252e150d))

## [1.1.0](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/compare/1.0.0...1.1.0) (2023-01-02)

### Features

- documentation for url-loader deactivation ([#4](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/issues/4)) ([8b8fe83](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/commit/8b8fe836550095730920cc74dc5558513ca5a1a8))

## [1.0.0](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/compare/0.2.2...1.0.0) (2022-12-26)

### ⚠ BREAKING CHANGES

- correct API

### Bug Fixes

- API now standard ([#2](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/issues/2)) ([61eae7c](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/commit/61eae7cccc97e230dec1abbef74525fa91cfb91b))

### [0.2.2](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/compare/0.2.1...0.2.2) (2022-12-26)

### [0.2.1](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/compare/0.2.0...0.2.1) (2022-12-25)

## [0.2.0](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/compare/0.1.2...0.2.0) (2022-12-25)

### Features

- make dependencies, peerDependencies ([016e235](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/commit/016e23580807f9662a8320018b78357ef9e6fdab))

### [0.1.2](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/compare/0.1.1...0.1.2) (2022-12-24)

### Bug Fixes

- build prior to release and ignore src when publishing ([a2bebf2](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/commit/a2bebf2bc5f4eed62a47f8f83e1bbd07c91e983e))

### [0.1.1](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/compare/0.1.0...0.1.1) (2022-12-24)

### Bug Fixes

- workflows ([98d06ec](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/commit/98d06ecb5025bc34c68f1f2937ceed0a2520fb7e))

## 0.1.0 (2022-12-24)

### Features

- initial version ([f8e03cd](https://github.com/johnnyreilly/rehype-cloudinary-docusaurus/commit/f8e03cd36977100f73f8c4f50efbf70110dc50cc))


src/plugin.test.ts
import type { Node } from "unist";
import { describe, expect, it } from "vitest";

import { imageCloudinaryRehypeVisitorFactory } from "./plugin.js";

describe("imageCloudinaryRehypeVisitorFactory", () => {
	it("transforms mdxJsxTextElement nodes", () => {
		const imageCloudinaryRehypeVisitor = imageCloudinaryRehypeVisitorFactory({
			cloudName: "demo",
			baseUrl: "https://johnnyreilly.com",
		});
		const node = {
			type: "mdxJsxTextElement",
			name: "img",
			attributes: [
				{
					type: "mdxJsxAttribute",
					name: "alt",
					value:
						"title image reading &quot;Azure Container Apps, Bicep, managed certificates and custom domains&quot; with the Azure Container App logos",
				},
				{
					type: "mdxJsxAttribute",
					name: "src",
					value: {
						type: "mdxJsxAttributeValueExpression",
						value:
							'require("!/home/john/code/github/blog.johnnyreilly.com/blog-website/node_modules/url-loader/dist/cjs.js?limit=10000&name=assets/images/[name]-[contenthash].[ext]&fallback=/home/john/code/github/blog.johnnyreilly.com/blog-website/node_modules/file-loader/dist/cjs.js!./screenshot-azure-portal-bring-your-own-certificates.webp").default',
						data: [],
					},
				},
				{ type: "mdxJsxAttribute", name: "width", value: "800" },
				{ type: "mdxJsxAttribute", name: "height", value: "450" },
			],
			children: [],
		} as const;

		imageCloudinaryRehypeVisitor(node as Node, undefined, undefined);

		expect(node.attributes[1].value.value).toBe(
			'`https://res.cloudinary.com/demo/image/fetch/f_auto,q_auto,w_auto,dpr_auto/https://johnnyreilly.com${require("!/home/john/code/github/blog.johnnyreilly.com/blog-website/node_modules/url-loader/dist/cjs.js?limit=10000&name=assets/images/[name]-[contenthash].[ext]&fallback=/home/john/code/github/blog.johnnyreilly.com/blog-website/node_modules/file-loader/dist/cjs.js!./screenshot-azure-portal-bring-your-own-certificates.webp").default}`'
		);
	});
});


src/plugin.ts
import * as acorn from "acorn";
import { fromMarkdown } from "mdast-util-from-markdown";
import { mdxJsxFromMarkdown, mdxJsxToMarkdown } from "mdast-util-mdx-jsx";
import { toMarkdown } from "mdast-util-to-markdown";
import { type Options, mdxJsx } from "micromark-extension-mdx-jsx";
import type { Transformer } from "unified";
import type { Node } from "unist";
import { type Visitor, visit } from "unist-util-visit";

import type { ImgOrJsxNodeData } from "./types.js";

interface Params {
	baseUrl: string;
	cloudName: string;
}

/**
 * Create a rehype plugin that will replace image URLs with Cloudinary URLs
 * @param {*} cloudName your Cloudinary’s cloud name eg demo
 * @param {*} baseUrl the base URL of your website eg https://blog.johnnyreilly.com - should not include a trailing slash, will likely be the same as the config.url in your docusaurus.config.js
 * @returns rehype plugin that will replace image URLs with Cloudinary URLs
 */
export function imageCloudinaryRehypePlugin({
	cloudName,
	baseUrl,
}: Params): Transformer {
	const imageCloudinaryRehypeVisitor = imageCloudinaryRehypeVisitorFactory({
		cloudName,
		baseUrl,
	});

	return (tree) => {
		visit(tree, ["mdxJsxTextElement"], imageCloudinaryRehypeVisitor);
	};
}

/**
 * Create a rehype visitor that will replace image URLs with Cloudinary URLs - exposed for testing purposes
 * @param {*} cloudName your Cloudinary’s cloud name eg demo
 * @param {*} baseUrl the base URL of your website eg https://blog.johnnyreilly.com - should not include a trailing slash, will likely be the same as the config.url in your docusaurus.config.js
 * @returns rehype plugin that will replace image URLs with Cloudinary URLs
 */
export function imageCloudinaryRehypeVisitorFactory({
	cloudName,
	baseUrl,
}: Params): Visitor<Node> {
	const srcRegex = / src=\{(.*)\}/;
	return function imageCloudinaryRehypeVisitor(node) {
		const imgWithAttributes = node as ImgOrJsxNodeData;
		if (
			imgWithAttributes.type === "mdxJsxTextElement" &&
			imgWithAttributes.name === "img"
		) {
			// handles nodes like this:
			// {
			//   type: 'mdxJsxTextElement',
			//   name: 'img',
			//   attributes: [
			//     {
			//       type: 'mdxJsxAttribute',
			//       name: 'alt',
			//       value: 'title image reading &quot;Azure Container Apps, Bicep, managed certificates and custom domains&quot; with the Azure Container App logos'
			//     },
			//     {
			//       type: 'mdxJsxAttribute',
			//       name: 'src',
			//       value: {
			//         type: 'mdxJsxAttributeValueExpression',
			//         value: 'require("!/home/john/code/github/blog.johnnyreilly.com/blog-website/node_modules/url-loader/dist/cjs.js?limit=10000&name=assets/images/[name]-[contenthash].[ext]&fallback=/home/john/code/github/blog.johnnyreilly.com/blog-website/node_modules/file-loader/dist/cjs.js!./screenshot-azure-portal-bring-your-own-certificates.webp").default',
			//         data: [Object]
			//       }
			//     },
			//     { type: 'mdxJsxAttribute', name: 'width', value: '800' },
			//     { type: 'mdxJsxAttribute', name: 'height', value: '450' }
			//   ],
			//   children: []
			// }
			const srcIndex = imgWithAttributes.attributes.findIndex(
				(attr) => attr.name === "src"
			);
			const requireAttribute = imgWithAttributes.attributes[srcIndex].value;
			if (typeof requireAttribute !== "string") {
				const asMarkdown = toMarkdown(
					imgWithAttributes as Parameters<typeof toMarkdown>[0],
					{
						extensions: [mdxJsxToMarkdown()],
					}
				);

				// <img
				//    alt="screenshot of typescript playground saying &#39;ComponentThatReturnsANumber&#39; cannot be used as a JSX component. Its return type &#39;number&#39; is not a valid JSX element.(2786)"
				//    src={require("!/home/john/code/github/blog.johnnyreilly.com/blog-website/node_modules/url-loader/dist/cjs.js?limit=10000&name=assets/images/[name]-[contenthash].[ext]&fallback=/home/john/code/github/blog.johnnyreilly.com/blog-website/node_modules/file-loader/dist/cjs.js!./screenshot-typescript-playground.png").default}
				//    width="690" height="298" />

				const match = asMarkdown.match(srcRegex);
				if (match) {
					const urlOrRequire = match[1];
					// eslint-disable-next-line no-useless-escape
					const cloudinaryRequireString = `\`https://res.cloudinary.com/${cloudName}/image/fetch/f_auto,q_auto,w_auto,dpr_auto/${baseUrl}\$\{${urlOrRequire}\}\``;

					const newMarkdown = asMarkdown.replace(
						srcRegex,
						` src={${cloudinaryRequireString}}`
					);

					// <img
					//    alt="screenshot of typescript playground saying &#39;ComponentThatReturnsANumber&#39; cannot be used as a JSX component. Its return type &#39;number&#39; is not a valid JSX element.(2786)"
					//    src={`https://res.cloudinary.com/priou/image/fetch/f_auto,q_auto,w_auto,dpr_auto/https://johnnyreilly.com${require("!/home/john/code/github/blog.johnnyreilly.com/blog-website/node_modules/url-loader/dist/cjs.js?limit=10000&name=assets/images/[name]-[contenthash].[ext]&fallback=/home/john/code/github/blog.johnnyreilly.com/blog-website/node_modules/file-loader/dist/cjs.js!./screenshot-typescript-playground.png").default}`}
					//    width="690" height="298" />

					const tree = fromMarkdown(newMarkdown, {
						extensions: [
							mdxJsx({
								acorn: acorn as unknown as Options["acorn"],
								addResult: true,
							}),
						],
						mdastExtensions: [mdxJsxFromMarkdown()],
					});

					const newNode = tree.children[0] as ImgOrJsxNodeData;
					const newSrcAttributeIndex = newNode.attributes.findIndex(
						(attr) => attr.name === "src"
					);

					if (newSrcAttributeIndex !== -1) {
						imgWithAttributes.attributes[srcIndex] =
							newNode.attributes[newSrcAttributeIndex];
					}
				}
			}
		}
	};
}
